gasprice/gasprice.go

```diff
@@ -0,0 +1,265 @@
+	// when starting node it will be some default value
+	LastPrice *big.Int
+	// IgnorePrice is the lowest price to take into consideration
+	// when collecting transactions
+	IgnorePrice *big.Int
+}
+
+// Backend is the interface representing blockchain and consensus backend
+type Backend interface {
+	GetBlockByHash(hash types.Hash, full bool) (*types.Block, bool)
+	Header() *types.Header
+	Config() *chain.Params
+}
+
+// GasStore interface is providing functions regarding gas and fees
+type GasStore interface {
+	// MaxPriorityFeePerGas calculates the priority fee needed for transaction to be included in a block
+	MaxPriorityFeePerGas() (*big.Int, error)
+}
+
+var _ GasStore = (*GasHelper)(nil)
+
+// GasHelper struct implements functions from the GasStore interface
+type GasHelper struct {
+	// numOfBlocksToCheck is the number of blocks to sample
+	numOfBlocksToCheck uint64
+	// pricePercentile is the sample percentile of transactions in a block
+	pricePercentile uint64
+	// sampleNumber is number of transactions sampled in a block
+	sampleNumber uint64
+	// maxPrice is the tip max price
+	maxPrice *big.Int
+	// lastPrice is the last price returned for maxPriorityFeePerGas
+	lastPrice *big.Int
+	// ignorePrice is the lowest price to take into consideration
+	// when collecting transactions
+	ignorePrice *big.Int
+	// backend is an abstraction of blockchain
+	backend Backend
+	// lastHeaderHash is the last header for which maxPriorityFeePerGas was returned
+	lastHeaderHash types.Hash
+
+	lock sync.Mutex
+}
+
+// NewGasHelper is the constructor function for GasHelper struct
+func NewGasHelper(config *Config, backend Backend) *GasHelper {
+	return &GasHelper{
+		numOfBlocksToCheck: config.NumOfBlocksToCheck,
+		pricePercentile:    config.PricePercentile,
+		sampleNumber:       config.SampleNumber,
+		ignorePrice:        config.IgnorePrice,
+		lastPrice:          config.LastPrice,
+		maxPrice:           config.MaxPrice,
+		backend:            backend,
+	}
+}
+
+// MaxPriorityFeePerGas calculates the priority fee needed for transaction to be included in a block
+// The function does following:
+//   - takes chain header
+//   - iterates for numOfBlocksToCheck from chain header to previous blocks
+//   - collects at most the sample number of sorted transactions in block
+//   - if not enough transactions were collected and their tips, go through some more blocks to get
+//     more accurate calculation
+//   - when enough transactions and their tips are collected, take the one that is in pricePercentile
+//   - if given price is larger then maxPrice then return the maxPrice
+func (g *GasHelper) MaxPriorityFeePerGas() (*big.Int, error) {
+	currentHeader := g.backend.Header()
+
+	currentBlock, found := g.backend.GetBlockByHash(currentHeader.Hash, true)
+	if !found {
+		return nil, fmt.Errorf(couldNotFoundBlockFormat, currentHeader.Number, currentHeader.Hash)
+	}
+
+	g.lock.Lock()
+	lastPrice := g.lastPrice
```

Comment:
Maybe we can use *lru.Cache instead of this optimization?

Reply:
Added `lru.cache` as suggested: https://github.com/0xPolygon/polygon-edge/pull/1629/commits/58c2de510e227952d09ff31151076a481279ae1f

Reply:
After looking at the code once more, we decided to drop the commit with `lru.cache` since we only take the last value calculated, so no need for a cache.